add deprecation warnings

rvagg opened the pull request:
See the changes in deb/src/_setup.sh and rpm/src/_setup.sh.

Print out warnings under certain conditions:

if you use a bare script, https://deb.nodesource.com/setup or https://rpm.nodesource.com/setup, we tell you that it's being deprecated and may soon be made inactive and that you should use a versioned script. This is done _at the begining of the script_ and you have to look at it for 10 seconds before it continues.

![screen shot 2016-06-18 at 4 59 42 pm](https://cloud.githubusercontent.com/assets/495647/16169564/d74dea02-3576-11e6-8d01-d5cd99a031c1.png)

If you use a v0.10 script it warns you of the Node deprecation in October and that you should upgrade. This is done at the _end_ of the script, just before the "run apt-get install nodejs" and "run yum install nodejs" bits, and waits for 5 seconds.

![screen shot 2016-06-18 at 5 00 06 pm](https://cloud.githubusercontent.com/assets/495647/16169568/e6b258de-3576-11e6-888d-620b9bbc3f84.png)

If you use a v0.12 script then it warns you of the Node deprecation at the end of the year and that you should upgrade. This is done at the _end_ of the script, just before the "run apt-get install nodejs" and "run yum install nodejs" bits, and waits for 3 seconds.

![screen shot 2016-06-18 at 5 00 16 pm](https://cloud.githubusercontent.com/assets/495647/16169569/f7ce67d4-3576-11e6-925f-813b9e4f2899.png)

Please review @retrohacker @chrislea @mweagle.

rvagg commented:
Oh yea, and if you use one of the bare `setup` scripts then you'll get both of these, one at the beginning and one at the end. It's possible that they'll look similar enough that they will ignore one of them, but if/when they switch to `setup_0.10` then they'll continue to get a warning anyway.

chrislea commented:
Too tired to actually sanity check ATM, but for the record, I <3 the sentiment here. :)

retrohacker commented:
@rvagg this PR replaces https://github.com/nodesource/distributions/pull/293 ?

I love the end result. I'll be able to review the shell script later this afternoon.

retrohacker commented on deb/setup:
```diff
@@ -35,9 +70,100 @@ exec_cmd() {
 
+node_deprecation_warning() {
+    if [ "XNode.js v0.10" == "XNode.js v0.10" ]; then
+        print_bold \
```
Is this function really helping all that much? We are littering `text` with style already... It looks like that function could be moved down into `node_deprecation_warning` by simply changing `print_bold` to 2 `echo`s inside the `if` statement.

retrohacker commented on deb/setup:
```diff
@@ -35,9 +70,100 @@ exec_cmd() {
+"
+
+        echo
+        echo "Continuing in 3 seconds ..."
```
Why 3 seconds here but 5 above?

rvagg replied:
cause 0.12 migration is less urgent, it ends at the end of 2016 whereas 0.10 ends at october .. just applying less pain than for v0.10 although it does seem a tiny bit arbitrary

retrohacker commented on deb/setup:
```diff
@@ -35,9 +70,100 @@ exec_cmd() {
+    elif [ "XNode.js v0.10" == "XNode.js v0.12" ]; then
+        print_bold \
+"                     NODE.JS v0.12 DEPRECATION WARNING                      " "\
+Node.js v0.10 will cease to be actively supported ${bold}at the end of 2016${normal}.
```
`Node.js v0.10` => `Node.js v0.12`

retrohacker replied:
`at the end of 2016`

We included a month 0.10 but were less specific here. Was that by design?

rvagg replied:
the "v0.10" here is cause I used `{{name}}` in the build script, this message isn't displayed to users cause of the `elif` condition but I'll change it anyway cause it'll be confusing to anyone viewing the script.

"end of 2016" is specific, it could say "end of december 2016" but it's really the end of the year

retrohacker commented on rpm/src/_setup.sh:
```diff
@@ -34,8 +69,100 @@ exec_cmd() {
 }
 
+node_deprecation_warning() {
+    if [ "X{{name}}" == "XNode.js v0.10" ]; then
```
This makes sense, in this file, but is super confusing in the other files. Something tells me we are going to get quite a few issues asking about this. Perhaps make a variable at the top of this script w/ `NAME={{name}}` and use that var here in the generated scripts?

Also, why are we putting an `X` in front?

rvagg replied:
good call re `NAME`, I like that idea

re `X`, that's just good practice in bash string comparison when doing variable expansion so you don't accidentally expose risks of execution or dodgy expansion, put something fixed at the front that can't be messed with. Not really important here but a good habit IMO.

rvagg commented:
As per @retrohacker's comments, in the latest commit I've moved the variables to the top and reused them as `${foo}` rather than `{{foo}}` substitutions so it's less confusing to people casually viewing the scripts. I've also removed the overuse of `{{name}}` in the deprecation text cause it's confusing if you're looking at the script.

retrohacker commented:
This is good to go by me.

Only concern with a change like this is how it affects automated tooling and `test -t` addresses this :smile:.

Moving forward, how do we signal to apps like chef and ansible that the script they are running is reaching EOL?

rvagg commented:
> Moving forward, how do we signal to apps like chef and ansible that the script they are running is reaching EOL?

That's a very good question ... one for which I don't have a good answer. I hope that the `sleep` sends a bit of a signal, which we can ramp up as we get closer to EOL. We're not going to discontinue serving v0.10 repos any time soon but perhaps some time after EOL, maybe 6 or 12 months, we start doing an `exit 1` from the script. We'll have download stats to give us some context at least.

Ultimately it's up to the user, there's only so much prodding we can do.

retrohacker commented:
`sleep` shouldn't get triggered by this script on ansible, `test -t` will keep that from triggering over ssh from my understanding. (ssh uses pipes under the hood I think, not a tty).

retrohacker commented:
Actually, that might be a decent approach. Only `sleep` on tty, otherwise just put it in their logs as `stderr` and hope that people are checking their logs for automated deploys.

I would guess that returning `1` will break a lot of automated tooling, that might be the last resort.

rvagg commented:
The `test -t 1` is only around the ANSI code stuff, if it's not a tty it'll fail and all of those `${red}` and `${bold}` things will be blank but they will still print and it'll still sleep.

rvagg commented:
updated with warnings for all 3 io.js versions and Node v5

![screen shot 2016-06-30 at 2 45 08 pm](https://cloud.githubusercontent.com/assets/495647/16477021/5efc8e3c-3ed1-11e6-8f12-898d15fa969e.png)

Merging and deploying, further updates will have to be in new PRs
